soc/intel/intel_adsp/tools/cavstool.py

```diff
@@ -666,14 +667,24 @@ def win_read(base, start, length):
         log.error("bar4_mmap.size()=%d", bar4_mmap.size())
         raise ie
 
-def win_hdr(base):
-    return struct.unpack("<IIII", win_read(base, 0, 16))
+def winstream_reg_hdr(base):
+    hdr = Regs(bar4_mem + base)
+    hdr.WLEN  = 0x00
+    hdr.START = 0x04
+    hdr.END   = 0x08
+    hdr.SEQ   = 0x0c
+    hdr.freeze()
+    return hdr
+
+def win_hdr(hdr):
+    return ( hdr.WLEN, hdr.START, hdr.END, hdr.SEQ )
```

Comment:
Not sure you really need this "read in a batch" API, and in the general case it can't work with lockless algorithms where the access order needs to be controlled (with winstream specifically it should be OK as IIRC the algorithm does the naive "check before and after the copy to see if it was clobbered otherwise retry" trick, which is order-independent).  Maybe just pass the Regs object around like the C struct that it actually is?

Reply:
Ack @andyross , this could be removed. This PR mostly touched the write implementation (where the ordering was more urgent to fix), but you are right similar cleanup can be done to read and drop the batch read.